File: sandbox/demo/child.py
import logging
import os

from sandbox.demo.common import ACK_BYTE, READY_BYTE
from sandbox.demo.config import Config
from sandbox.demo.diag import log_identity
from sandbox.demo.utils import flush_logs


def child(
    cfg: Config,
    p2c_r: int,
    c2p_w: int,
    host_uid: int,
    host_gid: int,
) -> int:
    logger = logging.getLogger("sandbox_demo.child")

    try:
        os.setgroups([])
    except Exception:
        logger.debug("os.setgroups([]) not permitted; continuing")

    logger.info("child starting before unshare()")
    log_identity(logger, "child (before unshare)", host_uid, host_gid)
    try:
        os.unshare(os.CLONE_NEWUSER | os.CLONE_NEWNET)
        logger.info("unshare(CLONE_NEWUSER) success")
    except OSError as e:
        logger.error(f"unshare(CLONE_NEWUSER) fail: {e}")
        return 1

    # tell parent that we are ready
    os.write(c2p_w, READY_BYTE)
    logger.debug("signaled ready to parent")

    # wait for ACK
    ack = os.read(p2c_r, 1)
    if ack != ACK_BYTE:
        logger.error("did not receive mapping ACK from parent; aborting")
        return 1

    # become root inside namespace
    try:
        os.setresgid(0, 0, 0)
    except PermissionError as e:
        logger.warning(f"setresgid(0) fail: {e}; continuing")
    try:
        os.setresuid(0, 0, 0)
    except PermissionError as e:
        logger.warning(f"setresgid(0) fail: {e}; continuing")

    logger.info("child is now root inside the user namespace")
    log_identity(logger, "child (after mapping + setuid(0))", host_uid, host_gid)

    # normal behavior: execute target program
    logger.info(f"exec: {cfg.command}")
    flush_logs()

    try:
        os.execvp(cfg.command[0], cfg.command)
    except Exception as e:
        logger.error(f"failed to execute target program: {e}")
        return 1

    # unreachable
    return 0
# Test change

File: sandbox/demo/child/child.go
//go:build linux

package child

import (
	"log"
	"os"
	"syscall"

	"sandbox/demo/common"
	"sandbox/demo/config"
	"sandbox/demo/diag"
	"sandbox/demo/network"
	"sandbox/demo/utils"
)

func Child(cfg config.Config, p2cR, c2pW uintptr, hostUID, hostGID int) int {
	_ = syscall.Setgroups([]int{})

	log.Println("[DEBUG] child starting before unshare()")
	diag.LogIdentity("child (before unshare)", hostUID, hostGID)

	flags := syscall.CLONE_NEWUSER
	if cfg.NetworkMode != network.Host {
		flags |= syscall.CLONE_NEWNET
	}

	if err := syscall.Unshare(flags); err != nil {
		log.Printf("[ERROR] unshare fail: %v", err)
		return 1
	}
	log.Printf("[DEBUG] unshare success (flags=%d)", flags)

	c2pWriter := os.NewFile(c2pW, "c2p_w")
	p2cReader := os.NewFile(p2cR, "p2c_r")

	_, _ = c2pWriter.Write([]byte{common.ReadyByte})
	c2pWriter.Close()
	log.Println("[DEBUG] signaled ready to parent")

	ackBuf := make([]byte, 1)
	if _, err := p2cReader.Read(ackBuf); err != nil || ackBuf[0] != common.AckByte {
		log.Println("[ERROR] did not receive mapping ACK from parent; aborting")
		return 1
	}
	p2cReader.Close()

	_ = syscall.Setresgid(0, 0, 0)
	_ = syscall.Setresuid(0, 0, 0)

	log.Println("[DEBUG] child is now root inside the user namespace")
	diag.LogIdentity("child (after mapping + setuid(0))", hostUID, hostGID)

	log.Printf("[DEBUG] exec: %v", cfg.Command)
	utils.FlushLogs()

	env := os.Environ()
	if err := syscall.Exec(cfg.Command[0], cfg.Command, env); err != nil {
		log.Printf("[ERROR] failed to execute target program: %v", err)
		return 1
	}

	return 0
}
